[SPARK-36374][SHUFFLE][DOC] Push-based shuffle high level user documentation

venkata91 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Document the push-based shuffle feature with a high level overview of the feature and corresponding configuration options for both shuffle server side as well as client side. This is how the changes to the doc looks on the browser ([img](https://user-images.githubusercontent.com/8871522/129231582-ad86ee2f-246f-4b42-9528-4ccd693e86d2.png))

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Helps users understand the feature

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Docs

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
N/A

venkata91 committed:
push-based shuffle documentation

venkata91 commented:
cc @mridulm @Victsm @Ngone51 @otterc @zhouyejoe Please take a look. Thanks.

SparkQA commented:
**[Test build #141975 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141975/testReport)** for PR 33615 at commit [`adc9dca`](https://github.com/apache/spark/commit/adc9dcaca36dd47006283c3e9289bc8926b1e891).

venkata91 commented:
I would like some suggestions/thoughts on where to place the push-based shuffle docs in the documentation website. After discussing with @Victsm adding it in the `Configuration` section seemed reasonable. Let me know if others have different thoughts.

SparkQA commented:
**[Test build #141975 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141975/testReport)** for PR 33615 at commit [`adc9dca`](https://github.com/apache/spark/commit/adc9dcaca36dd47006283c3e9289bc8926b1e891).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141975/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46486/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46486/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46486/

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+
+# Push-based shuffle overview
+
+Push-based shuffle is an improved shuffle architecture that optimizes the reliability and performance of the shuffle step in Spark. Complementing the existing shuffle mechanism, push-based shuffle takes a best-effort approach to push the shuffle blocks generated by the map tasks to remote shuffle services to be merged per shuffle partition. When the reduce tasks start running, they fetch a combin ...
```
`remote shuffle services` -> `remote external shuffle services`

Ngone51 replied:
`...into large sequential reads, which significantly reduce the overhead of disk I/O.`?

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+
+Push-based shuffle is an improved shuffle architecture that optimizes the reliability and performance of the shuffle step in Spark. Complementing the existing shuffle mechanism, push-based shuffle takes a best-effort approach to push the shuffle blocks generated by the map tasks to remote shuffle services to be merged per shuffle partition. When the reduce tasks start running, they fetch a combin ...
+
+<p> <b> Currently push-based shuffle is only supported for Spark on YARN. </b></p>
```
`.... supported for Spark on YARN with external shuffle service.`?

Ngone51 replied:
Shall we also recommend the appropriate scenario here? For example, the workload with the high load of disk I/O. AFAIK, the push-based shuffle doesn't perform well on small/fast queries yet.

venkata91 replied:
Ideally once we have https://issues.apache.org/jira/browse/SPARK-33701 (Adaptive shuffle merge finalization) fixed, small/fast queries should perform roughly the same with the existing shuffle mechanism. Do you think we should document it for now and once SPARK-33701 we can remove/reword it accordingly. Thoughts?

mridulm replied:
We are not targeting SPARK-33701 for 3.2 right ? Given that, let us document scenarios for 3.2 - for 3.3, SPARK-33701 would be an enhancement at which point 3.3 documentation would be enhanced to deal with the changes.

Victsm replied:
In that case, I think we should mention this improvement as a work in progress in the documentation to not misguide users.

venkata91 replied:
Makes sense. Added this info in the Overview section.

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+<table class="table">
+<tr><th>Property Name</th><th>Default</th><th>Meaning</th><th>Since Version</th></tr>
+<tr>
+  <td><code>spark.shuffle.server.mergedShuffleFileManagerImpl</code></td>
```
Is it able for users to plugin in the custom `MergedShuffleFileManager`?

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+    <code>org.apache.spark.network.shuffle.ExternalBlockHandler$NoOpMergedShuffleFileManager</code>
+  </td>
+  <td>
+    Class name of the implementation of MergedShuffleFileManager that manages push-based shuffle. This acts as a server side config to disable or enable push-based shuffle. By default, push-based shuffle is disabled at the server side, which does not perform push-based shuffle irrespective of the client side configuration. To turn on push-based shuffle at the server side, set the configuration to ...
```
> To turn on push-based shuffle at the server side, set the configuration to <code>org.apache.spark.network.shuffle.RemoteBlockPushResolver</code>

This sounds weird to me. I thought that the config `spark.shuffle.push.enabled` can enable it without extra configs. And I think we'd better set the config once for the better UX.

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.server.minChunkSizeInMergedShuffleFile</code></td>
+  <td><code>2m</code></td>
+  <td>
+    The minimum size of a chunk when dividing a merged shuffle file into multiple chunks during push-based shuffle. A merged shuffle file consists of multiple small shuffle blocks. Fetching the complete merged shuffle file in a single disk I/O increases the memory requirements for both the clients and the shuffle service. Instead, the shuffle service serves the merged file in <code> MB-sized chun ...
```
`...clients and the shuffle service.` -> `...clients and the external shuffle service.`

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.server.ioExceptionsThresholdDuringMerge</code></td>
+  <td><code>4</code></td>
+  <td>
+    The threshold for number of IOExceptions while merging shuffle blocks to a shuffle partition. When the number of IOExceptions while writing to the merged shuffle data/index/meta files exceed this threshold then the shuffle server will stop merging shuffle  blocks for this shuffle partition.
```
`shuffle  blocks` -> `shuffle blocks`

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+<table class="table">
+<tr><th>Property Name</th><th>Default</th><th>Meaning</th><th>Since Version</th></tr>
+<tr>
+  <td><code>spark.shuffle.server.mergedShuffleFileManagerImpl</code></td>
```
Shall we prefix `push` to all the server configs too?

venkata91 replied:
Ok. in that case we need to change the configs doc itself. Shall I just make those changes as well as part of this PR? One more thing I wanted to clarify is should the config `doc` and the documentation should reflect the same or it can be slightly different.

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.push.enabled</code></td>
+  <td><code>false</code></td>
+  <td>
+    Set to 'true' to enable push-based shuffle on the client side and this works in conjunction with the server side flag spark.shuffle.server.mergedShuffleFileManagerImpl which needs to be set with the appropriate org.apache.spark.network.shuffle.MergedShuffleFileManager implementation for push-based shuffle to be enabled
```
`be enabled` -> `be enabled.`

mridulm replied:
` and this works` -> `and works`

mridulm replied:
We can remove the suffix "which needs to be set with the appropriate org.apache.spark.network.shuffle.MergedShuffleFileManager implementation for push-based shuffle to be enabled"

It is documented for that flag above.

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.push.merge.finalize.timeout</code></td>
+  <td><code>10s</code></td>
+  <td>
+    Specify the amount of time DAGScheduler waits after all mappers finish for a given shuffle map stage before it starts sending merge finalize requests to remote shuffle services. This allows the shuffle services some extra time to merge as many blocks as possible.
```
Ideally, we should hide the internal knowledge from end-users. In this case, they may not know the `DAGScheduler`. Could you reword it?

mridulm replied:
`Specify the amount of time DAGScheduler waits after all mappers finish for a given shuffle map stage before it starts sending merge finalize requests to remote shuffle services. This allows the shuffle services some extra time to merge as many blocks as possible.` ->

`The amount of time driver waits, after all mappers have finished for a given shuffle map stage, before it sends merge finalize requests to remote shuffle services. This allows the shuffle services extra time to merge blocks.`

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.push.maxRetainedMergerLocations</code></td>
+  <td><code>500</code></td>
+  <td>
+    Maximum number of shuffle push merger locations cached for push based shuffle. Currently, shuffle push merger locations are nothing but external shuffle services which are responsible for handling pushed blocks and merging them and serving merged blocks for later shuffle fetch.
```
`...but the host of external shuffle services`?

mridulm replied:
`Maximum number of shuffle push merger locations cached for push based shuffle. Currently, shuffle push merger locations are nothing but external shuffle services which are responsible for handling pushed blocks and merging them and serving merged blocks for later shuffle fetch.` ->


`Maximum number of merger locations cached for push based shuffle. Currently, merger locations are hosts of external shuffle services responsible for handling pushed blocks, merging them and serving merged blocks for later shuffle fetch.`

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.push.numPushThreads</code></td>
+  <td><code>none</code></td>
+  <td>
+    Specify the number of threads in the block pusher pool. These threads assist in creating connections and pushing blocks to remote shuffle services. By default, the threadpool size is equal to the number of spark executor cores.
```
`By default, the threadpool size is equal to the number of spark executor cores.`

-> `If not set, the threadpool size is fallback to the value of spark.executor.cores`

mridulm replied:
Do we want to expose this config ?

Victsm replied:
This config might be a bit misleading. The threads in this thread pool does not really perform the work transferring the blocks to remote shuffle services. That is done by the Netty threads in executors.

Ngone51 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.push.maxBlockSizeToPush</code></td>
+  <td><code>1m</code></td>
+  <td>
+    The max size of an individual block to push to the remote shuffle services. Blocks larger than this threshold are not pushed to be merged remotely. These shuffle blocks will be fetched by the executors in the original manner.
```
Mention the side effect if the value is too big or too small.

mridulm replied:
Also, address similar for chunk size as well.

venkata91 replied:
Added the side effect info both for block size as well as chunk size confs respectively.

mridulm commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.push.mergersMinThresholdRatio</code></td>
+  <td><code>0.05</code></td>
+  <td>
+    The minimum number of shuffle merger locations required to enable push based shuffle for a stage. This is specified as a ratio of the number of partitions in the child stage. For example, a reduce stage which has 100 partitions and uses the default value 0.05 requires at least 5 unique merger locations to enable push based shuffle. Merger locations are currently defined as external shuffle se ...
```
`The minimum number of shuffle merger locations required to enable push based shuffle for a stage. This is specified as a ratio of the number of partitions in the child stage.` -> 

`Ratio used to compute the minimum number of shuffle merger locations required for a stage based on the number of partitions for the reducer stage`

mridulm replied:
We can remove `Merger locations are currently defined as external shuffle services.` ? That is already mentioned above.

mridulm reviewed (commented):
Took a pass through the PR - thanks for adding this @venkata91 !

venkata91 committed:
Address review comments from ngone51 and mridulm

SparkQA commented:
**[Test build #142051 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142051/testReport)** for PR 33615 at commit [`0f06bee`](https://github.com/apache/spark/commit/0f06beeee6e26e7ac1c80a8e82db8702434cf1ff).

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46562/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46562/

SparkQA commented:
**[Test build #142052 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142052/testReport)** for PR 33615 at commit [`484f4c5`](https://github.com/apache/spark/commit/484f4c5d619a1c94bd48a5144519e883dbb5dd06).

SparkQA commented:
**[Test build #142051 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142051/testReport)** for PR 33615 at commit [`0f06bee`](https://github.com/apache/spark/commit/0f06beeee6e26e7ac1c80a8e82db8702434cf1ff).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142051/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46563/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46563/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46563/

SparkQA commented:
**[Test build #142052 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142052/testReport)** for PR 33615 at commit [`484f4c5`](https://github.com/apache/spark/commit/484f4c5d619a1c94bd48a5144519e883dbb5dd06).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142052/

venkata91 committed:
change config in RemoteBlockPushResolverSuite

SparkQA commented:
**[Test build #142068 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142068/testReport)** for PR 33615 at commit [`fca6c0f`](https://github.com/apache/spark/commit/fca6c0fcae2b827373a22af0ef912a9e22cb55c9).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46578/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46578/

gengliangwang commented:
@venkata91 could you provide a screenshot for the documentation preview in PR description?

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46578/

gengliangwang commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,111 @@ The stage level scheduling feature allows users to specify task and executor res
+<tr>
+  <td><code>spark.shuffle.push.server.mergedShuffleFileManagerImpl</code></td>
+  <td>
+    <code>org.apache.spark.network.shuffle.ExternalBlockHandler$NoOpMergedShuffleFileManager</code>
```
It is hacky to contain `$` in the configuration value. Also, the config value is so long that it affects the readability of the doc:
![image](https://user-images.githubusercontent.com/1097932/128316212-16e704e4-b9c2-4ad5-80ec-1822387185a0.png)

gengliangwang replied:
Is it possible to move `NoOpMergedShuffleFileManager` under `org.apache.spark.network.shuffle`?

venkata91 replied:
Good suggestion! Would it cause issues while copy paste'ing the config for users?

gengliangwang replied:
> Would it cause issues while copy paste'ing the config for users

Yes it could be.

venkata91 replied:
Any suggestions?

gengliangwang replied:
Shall we move NoOpMergedShuffleFileManager under org.apache.spark.network.shuffle? So that the value becomes
`org.apache.spark.network.shuffle.NoOpMergedShuffleFileManager`

mridulm replied:
+CC @venkata91 Did you get a chance to raise a PR to refactor and pull this inner class out ?

venkata91 replied:
@mridulm Yeah I haven't tried that yet. But still the config key names are quite long, it would still not make the readability issue go away. Shouldn't this be handled at the CSS layer? Thoughts?

venkata91 replied:
Ok. I tried changing CSS for `td` to `word-wrap` works fine, but it also changes the other tables. We would still have the issue of `$` for the config value which is an issue. Let me file another PR to handle that.

venkata91 replied:
PR for pulling out `NoOpMergedShuffleFileManager` - https://github.com/apache/spark/pull/33688

gengliangwang commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,111 @@ The stage level scheduling feature allows users to specify task and executor res
+    <code>org.apache.spark.network.shuffle.ExternalBlockHandler$NoOpMergedShuffleFileManager</code>
+  </td>
+  <td>
+    Class name of the implementation of MergedShuffleFileManager that manages push-based shuffle. This acts as a server side config to disable or enable push-based shuffle. By default, push-based shuffle is disabled at the server side.
```
Shall we list possible values of the option and tell users how to enable the feature on the server-side?

venkata91 replied:
Added the info to the doc.

SparkQA commented:
**[Test build #142068 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142068/testReport)** for PR 33615 at commit [`fca6c0f`](https://github.com/apache/spark/commit/fca6c0fcae2b827373a22af0ef912a9e22cb55c9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142068/

venkata91 committed:
Address more review comments

venkata91 commented on docs/configuration.md:
```diff
@@ -3134,3 +3134,119 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.push.maxBlockBatchSize</code></td>
+  <td><code>3m</code></td>
+  <td>
+    The max size of a batch of shuffle blocks to be grouped into a single push request.
```
Added that part to the documentation.

Victsm replied:
The relevant part is to avoid setting this lower than `spark.storage.memoryMapThreshold`, which would trigger memory mapping for loading shuffle blocks for each push request.

venkata91 commented:
> @venkata91 could you provide a screenshot for the documentation preview in PR description?

Added a pdf file, image file didn't come out well.

SparkQA commented:
**[Test build #142118 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142118/testReport)** for PR 33615 at commit [`0ba3a0b`](https://github.com/apache/spark/commit/0ba3a0b9742a08372e74d5bc7a5a48f42a4c44a1).

venkata91 commented:
> @venkata91 could you provide a screenshot for the documentation preview in PR description?
Added to the PR description.

SparkQA commented:
**[Test build #142118 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142118/testReport)** for PR 33615 at commit [`0ba3a0b`](https://github.com/apache/spark/commit/0ba3a0b9742a08372e74d5bc7a5a48f42a4c44a1).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142118/

venkata91 commented:
Addressed the review comments, please take a look. cc @Ngone51 @mridulm @gengliangwang

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46630/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46630/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46630/

SparkQA commented:
**[Test build #142139 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142139/testReport)** for PR 33615 at commit [`57770b3`](https://github.com/apache/spark/commit/57770b39734fef06a9e03f4d934c7c7c7e49b099).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46651/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46651/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46651/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142139/

venkata91 committed:
Address readability comment

venkata91 committed:
Merge branch 'master' into SPARK-36374

SparkQA commented:
**[Test build #142238 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142238/testReport)** for PR 33615 at commit [`2d40171`](https://github.com/apache/spark/commit/2d401711143d1c48b684a8df9f5f7dfb92469846).

venkata91 committed:
Merge branch 'master' into SPARK-36374

venkata91 committed:
Address review comments

venkata91 commented:
Updated the PR with the recent changes from the master and also addressed the other pending review comments. Gentle ping to take a look again cc @mridulm @gengliangwang @Ngone51 Thanks.

SparkQA commented:
**[Test build #142252 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142252/testReport)** for PR 33615 at commit [`2aa5e72`](https://github.com/apache/spark/commit/2aa5e72c090fad0ac2868ca482d47f43ccc48b9b).

SparkQA commented:
**[Test build #142252 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142252/testReport)** for PR 33615 at commit [`2aa5e72`](https://github.com/apache/spark/commit/2aa5e72c090fad0ac2868ca482d47f43ccc48b9b).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142252/

venkata91 commented:
> **[Test build #142252 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142252/testReport)** for PR 33615 at commit [`2aa5e72`](https://github.com/apache/spark/commit/2aa5e72c090fad0ac2868ca482d47f43ccc48b9b).
> 
> * This patch **fails to build**.
> * This patch merges cleanly.
> * This patch adds no public classes.

It seems to fail to build because of ```[error] ## Exception when compiling 491 sources to /home/jenkins/workspace/SparkPullRequestBuilder/sql/catalyst/target/scala-2.12/classes
[error] java.lang.StackOverflowError``` but I don't think this is related to the PR. Is this a known issue?

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46759/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46759/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46759/

venkata91 committed:
Merge branch 'master' into SPARK-36374

mridulm reviewed (approved):
The changes look good to me, other than the minor comments above.
Please take a look @Ngone51, @gengliangwang

venkata91 committed:
Address mridulm review comments

SparkQA commented:
**[Test build #142344 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142344/testReport)** for PR 33615 at commit [`613aeae`](https://github.com/apache/spark/commit/613aeae54ff25f184f4ac2ce21f158490db43e5b).

venkata91 commented on docs/configuration.md:
```diff
@@ -3152,3 +3152,117 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td>3.2.0</td>
+</tr>
+<tr>
+  <td><code>spark.shuffle.push.server.ioExceptionsThresholdDuringMerge</code></td>
```
Makes sense, no need to expose this config.

mridulm replied:
Would be great to hear feedback from others as well ... it was just a thought :-)

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46853/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46853/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46853/

SparkQA commented:
**[Test build #142345 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142345/testReport)** for PR 33615 at commit [`26616b9`](https://github.com/apache/spark/commit/26616b9eed2af75378e6dda157b72c9c6983b978).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142345/

SparkQA commented:
**[Test build #142344 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142344/testReport)** for PR 33615 at commit [`613aeae`](https://github.com/apache/spark/commit/613aeae54ff25f184f4ac2ce21f158490db43e5b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class BlockPushNonFatalFailure extends RuntimeException `
  * `public class BlockPushReturnCode extends BlockTransferMessage `

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142344/

venkata91 commented:
cc @Victsm @otterc @zhouyejoe also to take a look once.

Ngone51 commented on docs/configuration.md:
```diff
@@ -3152,3 +3152,109 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td><code>spark.shuffle.push.merge.finalize.timeout</code></td>
+  <td><code>10s</code></td>
+  <td>
+    The amount of time driver waits in seconds, after all mappers have finished for a given shuffle map stage, before it sends merge finalize requests to remote shuffle services. This allows the shuffle services extra time to merge blocks.
```
"This allows the shuffle services extra time to merge blocks." -> "This gives the shuffle services extra time to merge blocks. Setting this too long could potentially lead to performance regression"

Ngone51 replied:
Also don't forget to update the corresponding `ConfigBuilder`.

venkata91 committed:
Address ngone51 review comments

venkata91 commented on docs/configuration.md:
```diff
@@ -3152,3 +3152,109 @@ The stage level scheduling feature allows users to specify task and executor res
+  <td>3.2.0</td>
+</tr>
+<tr>
+  <td><code>spark.shuffle.push.merge.finalize.timeout</code></td>
```
@Ngone51 2 reasons:
1. It can be slightly confusing to configure both `finalize.timeout` as well as `results.timeout` for users.
2. Also `results.timeout` is the upper bound, if the merge results arrive sooner than the upper bound (10s default) then driver just proceeds further scheduling the other stages. But that is not the case with `finalize.timeout` which is a fixed wait time before `finalize` is called on all the external shuffle services. This would also eventually become an upper bound once we address SPARK-33701. Hope that clarifies it a bit.
Let me know your thoughts. Thanks!

Ngone51 replied:
> It can be slightly confusing to configure both finalize.timeout as well as results.timeout for users.

I thought the same when I realized these two configs. Actually, I also got confused and spent some time to distinguish them.

And it sounds more like a system error when we exceed `results.timeout`, usually, it means a network issue. Even worse, the external shuffle service could be unresponsive even if we fall back to the original way to fetch the shuffle data.

Given all, shall we make it an `internal` conf? cc @mridulm @Victsm

mridulm replied:
I am fine with making it an internal conf - we can choose to expose it based on how SPARK-33701 evolves.

venkata91 replied:
Makes sense. Done.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46897/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46897/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46897/

SparkQA commented:
**[Test build #142390 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142390/testReport)** for PR 33615 at commit [`60563eb`](https://github.com/apache/spark/commit/60563eb62e8df8df9c82e821c17bdcf5778f8a27).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #142391 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142391/testReport)** for PR 33615 at commit [`ed8172f`](https://github.com/apache/spark/commit/ed8172f9418ade540c90aad8f7440c579a03afc1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142391/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142390/

venkata91 committed:
Address ngone51 review comments

SparkQA commented:
**[Test build #142435 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142435/testReport)** for PR 33615 at commit [`c112e2d`](https://github.com/apache/spark/commit/c112e2d6add021a3a279e19ffc48e5d80e11087e).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46942/

venkata91 committed:
Merge branch 'master' into SPARK-36374

SparkQA commented:
**[Test build #142441 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142441/testReport)** for PR 33615 at commit [`8585bb3`](https://github.com/apache/spark/commit/8585bb352b895d87b210e354355ae9f979c2bdbf).

SparkQA commented:
**[Test build #142441 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142441/testReport)** for PR 33615 at commit [`8585bb3`](https://github.com/apache/spark/commit/8585bb352b895d87b210e354355ae9f979c2bdbf).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142441/

mridulm commented:
There is a linter error for `TransportConf`, can you take a look at it @venkata91

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46942/

venkata91 committed:
Address lint issues

venkata91 commented:
> There is a linter error for `TransportConf`, can you take a look at it @venkata91

Fixed it. Thanks.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46942/

SparkQA commented:
Kubernetes integration test unable to build dist.

exiting with code: 1
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46947/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46947/

SparkQA commented:
**[Test build #142446 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142446/testReport)** for PR 33615 at commit [`1261828`](https://github.com/apache/spark/commit/12618284d95d7b1685c2c48c79bffcb618cf4523).

SparkQA commented:
**[Test build #142435 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142435/testReport)** for PR 33615 at commit [`c112e2d`](https://github.com/apache/spark/commit/c112e2d6add021a3a279e19ffc48e5d80e11087e).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142435/

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46952/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46952/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46952/

SparkQA commented:
**[Test build #142446 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/142446/testReport)** for PR 33615 at commit [`1261828`](https://github.com/apache/spark/commit/12618284d95d7b1685c2c48c79bffcb618cf4523).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/142446/

mridulm commented:
GA is passing, and the jenkins failure looks unrelated. Looks like the PR should be in a reasonable state to merge assuming no other changes are required.
Can you take a pass through the PR @Ngone51, @gengliangwang ? Thx !

gengliangwang commented on core/src/main/scala/org/apache/spark/internal/config/package.scala:
```diff
@@ -2097,86 +2097,87 @@ package object config {
         "org.apache.spark.network.shuffle.MergedShuffleFileManager implementation for push-based " +
         "shuffle to be enabled")
-      .version("3.1.0")
+      .version("3.2.0")
```
It seems that we have exposed this in 3.1.0(3.1.1) already: https://github.com/apache/spark/blob/branch-3.1/core/src/main/scala/org/apache/spark/internal/config/package.scala

Why do we change the version here? The configuration is only effective since 3.2?

venkata91 replied:
Yes, the feature was not ready in 3.1. It will be ready only in 3.2 which is why I changed it to 3.2 to keep it consistent.

mridulm commented:
Given RC timelines, will it be possible to take a pass @Ngone51, @gengliangwang ? I want to merge only after you are fine with the latest version. Thanks.

gengliangwang reviewed (approved):
LGTM, thanks for the work!

mridulm commented:
Merged to master, branch-3.2
+CC @gengliangwang 

Thanks for working on this @venkata91 !

Thanks for the reviews @Ngone51 , @gengliangwang, @Victsm :-)

Ngone51 commented:
Late LGTM, thanks for the work!
